model/waterfall.go

```diff
@@ -48,6 +49,68 @@ type WaterfallOptions struct {
 	MaxOrder   int      `bson:"-" json:"-"`
 	MinOrder   int      `bson:"-" json:"-"`
 	Requesters []string `bson:"-" json:"-"`
+	Variants   []string `bson:"-" json:"-"`
+}
+
+// Older versions don't have their build display names saved in the version document.
+// For those missing it, look up their build details.
+// TODO DEVPROD-15118: This function can be removed after 7 February 2026, when the version TTL index applies and all versions include build display names.
+func getBuildDisplayNames(match bson.M) bson.M {
+	match[bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusDisplayNameKey)] = bson.M{"$exists": false}
+	return bson.M{
+		"$unionWith": bson.M{
+			"coll": VersionCollection,
+			"pipeline": []bson.M{
+				bson.M{"$match": match},
+				bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}},
+				bson.M{"$limit": MaxWaterfallVersionLimit},
+				bson.M{
+					"$lookup": bson.M{
+						"from":         build.Collection,
+						"localField":   bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusIdKey),
+						"foreignField": build.IdKey,
+						"as":           VersionBuildVariantsKey,
+						"pipeline": []bson.M{
+							bson.M{
+								"$project": bson.M{
+									build.DisplayNameKey:    1,
+									VersionBuildStatusIdKey: build.IdKey,
+									build.BuildVariantKey:   1,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// This pipeline matches on versions that have a build with an ID or display name that matches variants
+func getBuildVariantFilterPipeline(variants []string, match bson.M) []bson.M {
+	pipeline := []bson.M{}
+	match[bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusDisplayNameKey)] = bson.M{"$exists": true}
+	pipeline = append(pipeline, bson.M{"$match": match})
+	matchCopy := bson.M{}
+	for key := range match {
+		matchCopy[key] = match[key]
+	}
+	pipeline = append(pipeline, getBuildDisplayNames(matchCopy))
+	pipeline = append(pipeline,
+		bson.M{
+			"$sort": bson.M{VersionRevisionOrderNumberKey: -1},
+		})
```

Comment:
nit. I think single line is slightly more readable
```suggestion
        pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
```